records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,113 @@
+
+    def copy(self,
+             loc: BaseDirectoryUrl,
+             other_loc: BaseDirectoryUrl) -> bool:
+        """Copy from one directory to another using optimized means.
+
+        :return: True if the copy was performed, or False if no
+        optimized means were possible."""
+        # TODO: Can I move some of this logic into the other classes?
+        if loc.scheme == 's3' and other_loc.scheme == 'file':
+            from records_mover.url.filesystem import FilesystemDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(loc, S3DirectoryUrl)
+            assert isinstance(other_loc, FilesystemDirectoryUrl)
+            return self._s3_copy_via_aws_cli.copy(loc, other_loc)
+        elif loc.scheme == 's3' and other_loc.scheme == 'gs':
+            from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(loc, S3DirectoryUrl)
+            assert isinstance(other_loc, GCSDirectoryUrl)
+            return self._gcp_data_transfer.copy(loc, other_loc)
+        else:
+            logger.info(f"No strategy to optimize copy from {loc} to {other_loc}")
+            return False
+
+    @contextmanager
+    def optimize_temp_second_location(self,
+                                      permanent_first_loc: BaseDirectoryUrl,
+                                      temp_second_loc: BaseDirectoryUrl) ->\
+            Iterator[BaseDirectoryUrl]:
+        # TODO: Can I move some of this logic into the other classes?
+        if permanent_first_loc.scheme == 's3' and temp_second_loc.scheme == 'gs':
+            from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(permanent_first_loc, S3DirectoryUrl)
+            assert isinstance(temp_second_loc, GCSDirectoryUrl)
+
+            with self._gcp_data_transfer.\
+                optimize_temp_second_location(permanent_first_loc,
+                                              temp_second_loc) as\
+                    optimized_second_loc:
+                logger.info(f"Optimized bucket location: {optimized_second_loc}")
+                yield optimized_second_loc
+        elif permanent_first_loc.scheme == temp_second_loc.scheme:
+            # Let's use the same location!
+            yield permanent_first_loc
+        else:
+            #
+            # No optimizations match
+            #
+            yield temp_second_loc
+
+    @contextmanager
+    def optimize_temp_locations(self,
+                                temp_first_loc: BaseDirectoryUrl,
+                                temp_second_loc: BaseDirectoryUrl) ->\
+            Iterator[Tuple[BaseDirectoryUrl, BaseDirectoryUrl]]:
+        # TODO: Can I move some of this logic into the other classes?
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker